Custom currency

gDanix opened the pull request:
Implementation of the user-defined currency.

Should close #2991

gDanix commented:
I don't know why fails build on AppVeyor. It seems that AppVeyor isn't building new custom-currency.c file.

Nubbie commented on data/language/en-GB.txt:
```diff
@@ -4188,6 +4188,16 @@ STR_5876    :{SMALLFONT}{BLACK}The engine to use for drawing the game.
+STR_5880    :Custom currency
+STR_5881    :Custom currency configuration
+STR_5882    :{WINDOW_COLOUR_2}Exchange rate: 
+STR_5883    :{WINDOW_COLOUR_2}is equivalent to {COMMA2DP32} GBP
```
Explain GBP (Add the symbol)
E.g. 
{WINDOW_COLOUR_2}is equivalent to {COMMA2DP32} GBP (£)

As it's referenced as  ``Pounds (£)`` In-game

janisozaur commented:
VS & Xcode explicitly specify files in their respective projects

gDanix commented:
@janisozaur can I generate that files with cmake?

Gymnasiast commented:
The settings windows says `(amount) is equivalent to 0.10 GBP`. This should read `(amount) is equivalent to 10 GBP`.

gDanix commented:
@Gymnasiast in currency.c, GBP is defined with a rate equal to 10 monetary units. This means that every monetary unit is equivalent to 1/10 GBP. Because we are defining the custom currency relatively to the monetary unit rather than to the GBP, that's why I show 0.10 GBP.

As you can see [here](https://github.com/OpenRCT2/OpenRCT2/pull/3952/commits/eabbe3eae21ded5df054373c54a72d21b2581c13#diff-dd629d2992492dc0edd61eb5495d25f3R233), I compute how many GBP are equivalent to 1 monetary unit. (I wrote 100/CurrencyDes... rather than 1/CurrencyDes... because [format_currency()](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/localisation/localisation.c#L453) and [format_currency_2dp()](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/localisation/localisation.c#L508) divides the value by 100, letting the value to actually be an integer representing a fixed point value (2 decimals position))

(Sorry if I didn't explained well!)

Gymnasiast commented:
Except it isn't true, as actually trying it will show you. Set the spinner to 1 Ctm and you'll see that all amounts are divided by 10: a park that has 10,000 GBP in cash will now have 1,000 Ctm.
So 1,000 Ctm = 10,000 GBP, thus
1 Ctm = 10 GBP.

Goddesen commented:
Then rather than comparing the value to £, maybe just write "units" instead? Comparing to pennies is just jarring.

Gymnasiast commented:
No, just put the correct amount in the window!

gDanix commented:
@Gymnasiast you're right!!

What a mess.... T_T

gDanix commented:
@Gymnasiast Thank you!!

Gymnasiast commented:
Good! :+1: 
Two small things:
- Make it default to 10 Ctm, rather than 1. (Since 10 Ctm = 10 GBP, I consider 10 Ctm the most useful default).
- Drop the two decimals, they no longer serve a purpose and `10 GBP` is more legible.

gDanix commented:
Done! :laughing:

marijnvdwerf commented on src/windows/custom_currency.c:
```diff
@@ -0,0 +1,287 @@
+
+/**
+ *
+ *  rct2: 0x0066D321
```
This is the address of the original method, but that's probably not true as this is a new feature

gDanix replied:
Right. I left them because copy/pasted from about.c window. Should I remove them?

marijnvdwerf commented on src/windows/custom_currency.c:
```diff
@@ -0,0 +1,287 @@
+#include "dropdown.h"
+
+enum WINDOW_CUSTOM_CURRENCY_WIDGET_IDX {
+    WIDX_BACKGROUND,
```
Other C files are indented with tabs.

gDanix replied:
The wiki (Coding style page) specifies that the indent should be done with spaces, 4 characters width, that's why I converted tabs into spaces.

Goddesen replied:
The coding style page mainly concerns new `C++` files. For C files, try to go with what is already established.

gDanix replied:
The [Coding style](https://github.com/OpenRCT2/OpenRCT2/wiki/Coding-style) page of teh wiki has been edited recently (June, 2), maybe that's causing the confusion.

It says that following these coding standards is not a priority. I considered replacing tabs by spaces important because spaces have a fixed width (tabs, depending on the editor configuration, can leave very much blank space in the code).

I don't think these changes are difficult to read but, anyway, if this is considered an important thing, I can change that, no problem.

Gymnasiast replied:
Following coding standards _is_ a priority in my book, but the wiki page starts off with saying it isn't. But de facto they are, we have frequently asked people to adhere. Clearly this needs discussion about those and about their enforcement, but not in this PR. 

Follow the standards in that page, but use tabs to indent (with a width of 4 spaces) for consistency with other C files.

gDanix replied:
Ok, I'll delete those addresses commented too because of what @marijnvdwerf pointed to.

Gymnasiast replied:
Yes, that is correct.

marijnvdwerf commented:
Not sure what's the desired spacing, but it's inconsistent in the PR itself at the moment.
`if(` vs `if (` and `a+1` vs `a + `1.

gDanix commented:
I'm not very strict with spaces, but if it's an important thing, I can review the PR and try to make the changes more consistent

Gymnasiast commented:
Oh, one more thing (sorry): could you invalidate the finance window upon changing anything? At the moment you have to move the finance window to update its contents.

IntelOrca commented:
Its better to just do a screen invalidation if you change something that can be present in lots of windows.

gDanix commented:
Sorry for taking so much time to fix this, I've been busy with university stuff.

I have refactored my code to follow the conding styles and rebased to fix a conflict that appeared these days (a previous commit added STR_5880 and STR_5881).

Gymnasiast commented:
I'll try to test it tonight.

Gymnasiast commented on src/windows/options.c:
```diff
@@ -914,16 +914,27 @@ static void window_options_mousedown(int widgetIndex, rct_window*w, rct_widget*
 
-			dropdown_set_checked(gConfigGeneral.currency_format, true);
+			if(gConfigGeneral.currency_format == CURRENCY_CUSTOM){
+				dropdown_set_checked(gConfigGeneral.currency_format+1, true);
```
There should be spaces around the `+`. That also applies to other occurrences of it, and to the `-` sign as well.

Gymnasiast commented:
I noticed that if you set the exchange rate to, for example, 11 Ctm to £10 it will stop displaying currencies with two decimal places. That should happen from 100 rather than 11, as it will cause £2 and £2.10 to appear exactly the same.

I also left a note about formatting.

gDanix commented:
I think I found why this happened. [format_currency_2dp()](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/localisation/localisation.c#L507) is actually dropping decimals when the `rate` is greater than 10, rather than checking the `value` being formatted.

Gymnasiast commented:
As far as I'm concerned, this can be merged. @IntelOrca @duncanspumpkin

gDanix commented:
What I finally did was to raise `rate` from 10 to 100, but this solution can be discussed, because currencies whose exchange rate isn't multiple of 10, will still need decimals.

The best solution to avoid headaches is to set the exchange rate as a multiple of 10.

IntelOrca commented on src/openrct2.c:
```diff
@@ -323,6 +323,11 @@ void openrct2_launch()
+		// Set custom currency properties
+		CurrencyDescriptors[CURRENCY_CUSTOM].rate = gConfigGeneral.custom_currency_rate;
+		CurrencyDescriptors[CURRENCY_CUSTOM].affix_unicode = gConfigGeneral.custom_currency_affix;
+		strncpy(CurrencyDescriptors[CURRENCY_CUSTOM].symbol_unicode, gConfigGeneral.custom_currency_symbol, CURRENCY_SYMBOL_MAX_SIZE);
```
This isn't a suitable place to put this code, it should be done in config load.

gDanix replied:
You're right, I couldn't find the config load functions when I added these lines. I'm now working on this.

gDanix replied:
It's acceptable to move that code into a new currency.c function and call it from openrct2_initialise()?

Gymnasiast commented:
I think your solution was perfectly adequate.

IntelOrca commented:
@gDanix call it from `config_open_default` because it needs to always be called if the game is re-loading the config.
